source/ring_buffer.c

```diff
@@ -268,3 +268,55 @@ bool aws_ring_buffer_buf_belongs_to_pool(const struct aws_ring_buffer *ring_buff
     AWS_POSTCONDITION(aws_byte_buf_is_valid(buf));
     return rval;
 }
+
+/* Ring buffer allocator implementation */
+static void *s_ring_buffer_mem_acquire(struct aws_allocator *allocator, size_t size) {
+    struct aws_ring_buffer *buffer = allocator->impl;
+    struct aws_byte_buf buf;
+    AWS_ZERO_STRUCT(buf);
+    /* allocate extra space for the size */
+    if (aws_ring_buffer_acquire(buffer, size + sizeof(size_t), &buf)) {
+        return NULL;
+    }
+    /* store the size ahead of the allocation */
+    *((size_t *)buf.buffer) = buf.capacity;
+    return buf.buffer + sizeof(size_t);
+}
+
+static void s_ring_buffer_mem_release(struct aws_allocator *allocator, void *ptr) {
+    /* back up to where the size is stored */
+    const void *addr = ((uint8_t *)ptr - sizeof(size_t));
+    const size_t size = *((size_t *)addr);
+
+    struct aws_byte_buf buf = aws_byte_buf_from_array(addr, size);
+    buf.allocator = allocator;
+
+    struct aws_ring_buffer *buffer = allocator->impl;
+    aws_ring_buffer_release(buffer, &buf);
+}
+
+static void *s_ring_buffer_mem_calloc(struct aws_allocator *allocator, size_t num, size_t size) {
+    void *mem = s_ring_buffer_mem_acquire(allocator, num * size);
+    if (!mem) {
+        return NULL;
+    }
+    memset(mem, 0, num * size);
+    return mem;
+}
+
+int aws_ring_buffer_allocator_init(struct aws_allocator *allocator, struct aws_ring_buffer *ring_buffer) {
+    if (allocator == NULL || ring_buffer == NULL) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    allocator->impl = ring_buffer;
+    allocator->mem_acquire = s_ring_buffer_mem_acquire;
+    allocator->mem_release = s_ring_buffer_mem_release;
+    allocator->mem_calloc = s_ring_buffer_mem_calloc;
```

Comment:
I worry here that realloc will accidentally cause an out of order allocation/free since we didn’t implement it the fallback will be used. Maybe we should assert in realloc?

Reply:
We don't define realloc, (it's null), so aws_mem_realloc() will assert for us.

Reply:
https://github.com/awslabs/aws-c-common/blob/master/source/allocator.c#L184

Reply:
NOPE, you were right. Fixed.